[CORE-10235] cluster_link: Implement DeleteShadowLink API

IoannisRP opened the pull request:
Implements: [CORE-10235](https://redpandadata.atlassian.net/browse/CORE-10235)

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

[CORE-10235]: https://redpandadata.atlassian.net/browse/CORE-10235?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

IoannisRP committed:
dt: factor our helper function in ShadowLinkBasicTests

IoannisRP committed:
cluster_link: add new error enum value

IoannisRP committed:
dt: add delete link funtion in ShadowLinkBasicTests

This api is still unimplemented in redpanda. It will be exposed in a later commit.

IoannisRP committed:
cluster_link: implement delete cluster link path

This implements all the routing and behavior needed for the DeleteShadowLink
API endpoint. It will be exposed in a later commit.

IoannisRP committed:
r/a/s/shadow_link: expose DeleteShadowLink API

michael-redpanda commented on src/v/cluster_link/errc.cc:
```diff
@@ -58,6 +58,8 @@ struct error_category final : public std::error_category {
         case errc::link_creation_failed:
             return "link creation failed";
+        case errc::link_has_mirrored_topics:
+            return "link has mirrored topics";
```
nit: `link has active shadow topics`

mirrored topics -> shadow topics, and "active" I Think is important as well

michael-redpanda commented on src/v/cluster_link/manager.cc:
```diff
@@ -208,6 +208,34 @@ result<chunked_vector<model::metadata>> manager::list_cluster_links() {
+        co_return cl_resp.assume_error();
+    }
+
+    const auto& mirror_topics = cl_resp.assume_value().state.mirror_topics;
+    if (!mirror_topics.empty()) {
+        co_return err_info(
+          errc::link_has_mirrored_topics,
+          fmt::format(
+            "Failed to delete cluster link with name '{}'. There are active "
+            "mirrored topics.",
+            name));
+    }
```
change: Not quite correct, you need to check the states of those mirror topics: https://github.com/redpanda-data/redpanda/blob/a08df8205b2a70d1ce0a59e374395a576903d09f/src/v/cluster_link/model/types.h#L39-L48

I'd say it's permitted to remove if the state != active or paused

michael-redpanda commented on tests/rptest/tests/cluster_linking_test_base.py:
```diff
@@ -176,6 +176,13 @@ def create_default_link_request(
+    def delete_link_request(
+        self, link_name: str
+    ) -> shadow_link_pb2.DeleteShadowLinkRequest:
+        req = shadow_link_pb2.DeleteShadowLinkRequest()
+        req.name = link_name
```
nit: you should be able to pass the name of the link to the constructor of `DeleteShadowLinkRequest`, so something like:
```python
req = shadow_link_pb2.DeleteShadowLinkRequest(name=link_name)
```

michael-redpanda commented on tests/rptest/tests/cluster_linking_test_base.py:
```diff
@@ -187,6 +194,13 @@ def create_link_with_request(
+        req = self.delete_link_request(link_name=link_name, *args, **kwargs)
+        return self.delete_link_with_request(req=req)
+
+    def delete_link_with_request(self, req: shadow_link_pb2.DeleteShadowLinkRequest):
+        return self.service_client.delete_shadow_link(req=req)
```
nit: have this return a `DeleteShadowLinkResponse` - even though it's empty just forward looking

IoannisRP replied:
it already did. Added the type hint to be more explicit.

michael-redpanda replied:
yeah I like those type hints... call me an old school C++ developer 😆

michael-redpanda reviewed (changes requested):
looks good, some changes requested

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/a28e97a1f377b28cb7c03a54c401472ba37a2d7f..ca8964d4de904fbedf728a478140cacbe5c2a477):
- rebased to dev and resolved merge conflict

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/ca8964d4de904fbedf728a478140cacbe5c2a477..9f32297e18a238a0dd472de1b6c02e0448649a86):
- renamed error to include "active shadow" topics
- fixed active topic check to verify status
- added type hints in the delete_link functions

michael-redpanda commented on src/v/cluster_link/manager.cc:
```diff
@@ -208,6 +208,49 @@ result<chunked_vector<model::metadata>> manager::list_cluster_links() {
+        }
+    };
+
+    const auto mirror_topic_states = cl_resp.assume_value().state.mirror_topics
+                                     | std::views::values
+                                     | std::views::transform(
+                                       &model::mirror_topic_metadata::state);
+
+    if (std::ranges::any_of(mirror_topic_states, is_active)) {
+        co_return err_info(
+          errc::link_has_active_shadow_topics,
+          fmt::format(
+            "Failed to delete cluster link with name '{}'. There are active "
+            "shadow topics.",
+            name));
+    }
```
praise: 🔥

michael-redpanda commented:
Force push:

* rebased off of dev

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--b3b86b47-7857-47fe-8d3e-54e49cc3191a-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72377">build#72377</a></summary>


| test_class               | test_method             | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodePostRestartProbeTest | post_restart_probe_test | null             | integration | https://buildkite.com/redpanda/redpanda/builds/72377#019953fd-8dcb-4ff1-8ae3-8313cf6f1473 | FLAKY         | 20/21    | upstream reliability is '99.53051643192488'. current run reliability is '95.23809523809523'. drift is 4.29242 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->
